job_definitions/update_api_tokens.yml

```diff
@@ -0,0 +1,74 @@
+{% set stages = [{'name': 'Add new tokens', 'command': 'add-new'}, {'name': 'Remove old tokens', 'command': 'remove-old'}] %}
+{% set app_groups = [['search-api'], ['api'], ['admin-frontend', 'briefs-frontend', 'brief-responses-frontend', 'buyer-frontend', 'supplier-frontend', 'user-frontend']] %}
+---
+- job:
+    name: update-api-tokens
+    display-name: "Update API tokens"
+    project-type: pipeline
+    descript1ion: Update Data and Search API tokens for a given environment.
+    disabled: false
+    concurrent: false
+    parameters:
+      - choice:
+          name: STAGE
+          choices:
+            - preview
+            - staging
+            - production
+    pipeline:
+      script: |
+        def notify_slack(icon, status) {
+          build job: "notify-slack",
+          parameters: [
+            string(name: 'USERNAME', value: "update_api_tokens"),
+            string(name: 'ICON', value: icon),
+            string(name: 'JOB', value: "Update API tokens on ${STAGE}"),
+            string(name: 'CHANNEL', value: "#dm-release"),
+            text(name: 'STAGE', value: "${STAGE}"),
+            text(name: 'STATUS', value: status),
+            text(name: 'URL', value: "<${BUILD_URL}consoleFull|${BUILD_DISPLAY_NAME}>")
+          ]
+        }
+
+{% for stage in stages %}
+        stage("{{ stage.name }}") {
+{% if loop.index == 1 %}
+          currentBuild.displayName = "#${BUILD_NUMBER} - ${STAGE}"
+{% endif %}
+          node {
+            try {
+              sh('docker run -e GITHUB_ACCESS_TOKEN digitalmarketplace/scripts scripts/update-api-tokens.sh "{{ stage.command }}" "${STAGE}"')
+
+              notify_slack(':spinner:', '{{ stage.name }} - SUCCESS')
+
+            } catch(e) {
+              notify_slack(':kaboom:', '{{ stage.name }} - FAILED')
+              echo "Error caught"
+              currentBuild.result = 'FAILURE'
+              return
+            }
+          }
+        }
+
+        stage('Deploy') {
+          milestone()
+          input(message: "Do not continue until the Pull Request updating API tokens has been merged to master.")
+          milestone()
+
+{% for app_group in app_groups %}
```

Comment:
The docs seem to suggest that only stages within a `parallel` block will be run in parallel. This gives you three `parallel` blocks, but the first two only have one stage in them for api and search-api. It probably doesn't do any harm but could be confusing.

Also, what does having `"{{ app }}"` as the top level key give you? In the docs it's just got `stage` as the top level.

Reply:
Hmm, in the docs I looked at you need to give parallel a map, so the app name is just filling that need rather than anything more meaningful. 🤔

Reply:
I was looking here: https://jenkins.io/blog/2017/09/25/declarative-1/

Reply:
https://jenkins.io/doc/book/pipeline/jenkinsfile/#parallel-execution

Reply:
As for why three parallel blocks when the first two are redundant - you're right, it doesn't actually do anything in the case of the search-api and the api. Those will happen consecutively rather than at the same time, but that's what they need to do. It was just a slightly simpler template to group them and have them all run within parallel blocks.

Reply:
This is the alternative:


```{% for app_group in app_groups %}
{% if app_group|length == 1 %}
          stage("{{ app_group[0] }}") {
            build job: "release-app-paas", parameters: [
              string(name: "STAGE", value: "${STAGE}"),
              string(name: "APPLICATION_NAME", value: "{{ app_group[0] }}")
            ]
          }
{% elif app_group|length > 1 %}
          parallel(
{% for app in app_group %}
            "{{ app }}": {
              stage("{{ app }}") {
                build job: "release-app-paas", parameters: [
                  string(name: "STAGE", value: "${STAGE}"),
                  string(name: "APPLICATION_NAME", value: "{{ app }}")
                ]
              }
            },
{% endfor %}
          )
{% endif %}
{% endfor %}
        }```

Reply:
Gross. Just keep the original I think.